DBZ-4366 basic support for DBMS_LOB package

chanetd opened the pull request:
Make less assumptions about the order of SELECT_LOB_LOCATOR and
LOB_WRITE events when a LOB value is constructed. The current approach
also supports interleaved LOB creation (piecemeal creation of 2 large
LOB values in different rows or columns at the same time).

There is still no support for partial LOB manipulations (trimming a LOB,
appending a couple of bytes, overwriting just a couple of bytes in the
middle, ...). To properly support that, a wholly different approach is
needed.

Naros commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/infinispan/marshalling/LobWriteEventAdapter.java:
```diff
@@ -63,4 +65,26 @@ public LobWriteEvent factory(int eventType, String scn, String tableId, String r
+     * @param event the event instance, must not be {@code null}
+     * @return the offset of the data to be written for a LOB field
+     */
+    @ProtoField(number = 8, required = true)
```
Marking `getOffset` and `getLength` as `required` will mean that the code isn't backward compatible and older caches won't be loadable.  That's likely fine since this is all still incubating; so we'll need to include a migration note about this.

chanetd replied:
I can also make them optional, and add some logic to deal with missing values in TransactionCommitConsumer. That will complicate the code solely for the purpose of dealing with this transition case, though, so I'm not sure if adding the migration note isn't the more pragmatic solution.
Up to you - if you want me to deal with optional length/offset fields, I can figure out a way to fit that in.

Naros replied:
Nah; Infinispan support was added in 1.7 and it's still technically "incubating" so I think it's fine to require a re-snapshot.  I mostly made the note here in the PR so we don't lose sight of the migration note needed.  I'll add something to the issue right now so it's where we need it for release purposes.

Naros commented on debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleBlobDataTypesIT.java:
```diff
@@ -1401,6 +1401,242 @@ public void shouldStreamBlobWithUnavailableColumnValuePlaceholder() throws Excep
 
+    @Test
+    @FixFor("DBZ-4366")
+    public void shouldStreamBlobsWrittenInChunkedMode() throws Exception {
```
Could a similar test be included in `OracleClobDataTypesIT` for `CLOB` and `NCLOB` so we can verify and have test coverage for those in case there are any variants in behavior?

Naros commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/TransactionCommitConsumer.java:
```diff
@@ -109,233 +102,383 @@ public void accept(LogMinerEvent event) throws InterruptedException {
+            // the uncommon case: writing somewhere in the middle
+
+            // find the right spot to insert
+            ListIterator<LobFragment> iter = fragments.listIterator();
+            while (iter.hasNext()) {
+                LobFragment frag = iter.next();
+                if (fragment.offset < frag.end() && fragment.offset >= frag.offset) {
+                    // truncate frag and insert after
+                    frag.truncate(fragment.offset - frag.offset);
+                    iter.add(fragment);
+                    break;
+                }
+                if (frag.offset > fragment.offset) {
+                    // insert before; no need to truncate preceding fragment
+                    iter.previous();
+                    iter.add(fragment);
+                    break;
+                }
+            }
+
+            // are there any following fragments that are (partially) overwritten by the fragment we're inserting?
+            while (iter.hasNext()) {
+                LobFragment frag = iter.next();
+                if (frag.offset >= fragment.end()) { // we're done
+                    break;
+                }
+                if (frag.end() <= fragment.end()) { // remove entirely
+                    iter.remove();
+                }
+                else { // front-truncate
+                    frag.frontTruncate(frag.end() - fragment.end());
+                }
+            }
```
Have you measured the performance impact of this iteration at all?  Particularly when the size of the file/character data grows to a respectable size of 5MB, 10MB, 25MB?  Are there any optimizations that could be done here to avoid or minimize the iteration per fragment added?  I'm just concerned this could introduce performance problems when each fragment is writing 8k or 16k of data in a large 10MB or 25MB file.

chanetd replied:
It's the uncommon case (the expected case is all-appends, which is handled efficiently above), so I didn't spend too much time thinking about optimizing this, in the interest of keeping the code simple and understandable.
If you want, I can look into this.

Actually, I now see that I missed a corner case in this algorithm: when a new fragment overwrites the middle of an existing fragment (say, there is a fragment with offset 1000, length 1000 and we add a new fragment with offset 1400, length 100), the remainder of the original fragment (offsets 1500-2000) will be truncated erroneously. I'll go back to the drawing board to fix that, and perhaps add an additional test for this case.

Naros replied:
Sounds great.

Naros commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/TransactionCommitConsumer.java:
```diff
@@ -109,233 +102,383 @@ public void accept(LogMinerEvent event) throws InterruptedException {
+    private void acceptLobManipulationEvent(LogMinerEvent event) {
+        if (null == currentLobRowId || null == currentLobColumnName) {
+            // should only happen when we start streaming in the middle of a LOB transaction (DBZ-4367)
+            LOGGER.trace("got LOB manipulation event without preceding LOB selector; ignoring {} {}", event.getEventType(), event);
```
It's a small nit-pick but could we be consistent with the logging messages such that the first letter of the first word is capitalized?  There are multiple occurrences of this in this PR, so I've only pointed out this one.  Thanks!

Naros commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/TransactionCommitConsumer.java:
```diff
@@ -66,38 +71,26 @@
-            dispatchChangeEvent(lastEvent);
+        // sort the remaining events on increasing SCN before dispatch
+        List<DmlEvent> events = new ArrayList<DmlEvent>(rows.values());
+        Collections.sort(events, (a, b) -> a.getScn().compareTo(b.getScn()));
```
I don't think we should try to implement any behavior based on sorting by system change numbers.  Oracle is allowed to emit a certain number of events with the same SCN and therefore we ultimately should rely on the order based on how they've been read as the source of record for their order than doing this.  Is there then any concern with this sort?

chanetd replied:
The purpose of this sort was to make the records come out in a somewhat-predictable way. The implementation will typically emit all records from the `rows` map at the invocation of `TransactionCommitConsumer#close()` (because we hold on to the last "event batch" for each row until the TX is closed). That means that in a TX with lots of affected rows, the order in which they are emitted is probably totally unrelated to the order in which the events were read from LogMiner (the iteration order over the HashMap's value set will determine emission order). I thought the SCNs would be a good proxy for the "original" order, but I can make that original order explicit by adding in a counter in the row state, and sorting on that counter instead.

Naros replied:
I had the same inclination as well when I did some rework back in 1.4 or so and quickly found that applying any sort of order whether at the query level or based on objects in a collection caused problems.  In general, it's best to let LogMiner give us the order and we adhere to that.

Naros reviewed (commented):
I haven't pulled down this code directly @chanetd but I wanted to give you some quick feedback just based on a quick review of what I see.  I've added some comments inline.

gunnarmorling commented:
@Naros, I suppose this needs more baking still, right?

chanetd commented:
@Naros I pushed a new commit, addressing all of your comments:
- fixed log capitalization
- on `TransactionCommitConsumer#close()`, the remaining buffered events are now emitted in the order we ingested them
- added a CLOB integration test
- fixed the LOB construction error I spotted, and added unit tests for LobUnderConstruction
- made an effort to keep the number of fragments in a LobUnderConstruction small in case of out-of-order writes, to reduce the impact of the linear search. I could have introduced a more complex data structure instead, but I don't think this is a very common case, so I'm hesitant to complicate the code too much to accommodate it.

One thing I'm still unhappy about is the representation of BLOB fragments. In keeping with the original code, I use the `HEXTORAW('...')` representation, but when you start fusing and partially overwriting fragments, this is quite inconvenient. I think the cleanest solution would be to do away with `HEXTORAW('')` altogether, and rely on a flag to mark the contents as binary instead of textual, and do this across the board from TransactionCommitConsumer (or perhaps even from LogMinerEvent) to OracleValueConverter. What do you think?

Naros commented:
I can certainly see where the fusing/partial writing of fragments can be annoying with the `HEXTORAW('...')` function call wrapper.  In terms of an intermediate representation, I don't particularly have a preference for how it gets represented during this step as long as whatever we do keeps the overhead as low as possible.  

My only ask is that we are clear in the code/java docs about what we're doing and where this gets re-applied in the pipeline and that whatever we do is also consistent with Xstream.  As to where we reapply the function call wrapper, it might be best that we delay that until `TransactionCommitConsumer` when we're handing the event off to the core framework.  This prevents the need to apply the function wrapper during loading events from or serializing events to disk via Infinispan.

You agree?

chanetd commented:
As far as I can tell from reading the code, the XStream implementation doesn't use the `HEXTORAW` wrapper at all. It gets ChunkedColumnValues from the XStream API, and these are plain byte arrays under the hood.
So it might actually improve consistency with XStream to do away with HEXTORAW and BlobChunkList altogether, and simply pass a byte[] down into the core framework.

chanetd commented:
The newest version of this PR removes BlobChunkList and the HEXTORAW(...) encoding from LogMiner. It's `byte[]` all the way now for BLOBs.

Naros commented:
Hi @chanetd sorry this PR has lingered a bit; we had holiday breaks and what-not.  I'm going to take a look at this today and offer any last-minute feedback if any.  Thanks again for all the effort & your patience.

Naros commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/AbstractLogMinerEventProcessor.java:
```diff
@@ -626,7 +628,8 @@ protected void handleLobWrite(LogMinerEventRow row) {
 
         if (row.getRedoSql() != null) {
-            addToTransaction(row.getTransactionId(), row, () -> new LobWriteEvent(row, parseLobWriteSql(row.getRedoSql())));
+            final ParsedLobWriteSql parsed = parseLobWriteSql(row.getRedoSql());
+            addToTransaction(row.getTransactionId(), row, () -> new LobWriteEvent(row, parsed.data, parsed.offset, parsed.length));
```
Do we need to explicitly always invoke the parse on the `LOB_WRITE` event?  I specifically moved a number of these into the actual Lambda function instead so that if `addToTransaction` decides to discard the event that we don't unnecessarily parse the SQL text.  So unless there is a specific reason to do this always, I would suggest we adjust these lines to look like the following:

```suggestion
            addToTransaction(row.getTransactionId(), row, () -> {
              final ParsedLobWriteSql parsed = parseLobWriteSql(row.getRedoSql());
              return new LobWriteEvent(row, parsed.data, parsed.offset, parsed.length));
            });
```

chanetd replied:
done

Naros commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/TransactionCommitConsumer.java:
```diff
@@ -144,198 +254,338 @@ private void dispatchChangeEvent(LogMinerEvent event) throws InterruptedExceptio
-        final String columnName = lastSelectLobLocatorEvent.getColumnName();
-        final int columnIndex = getSelectLobLocatorColumnIndex();
+        void append(LobFragment other) {
+            assert other.offset >= end();
+            assert other.binary == binary;
```
I would prefer if we want to keep these sanity checks, lets turn them into legit DebeziumException raising call sites rather than utilizing `assert` semantics.

chanetd replied:
done

Naros reviewed (commented):
Hi @chanetd overall this looks really good.  I still need to give a test drive locally but I wanted to give you some immediate feedback; below are a few things I saw inline:

Naros commented:
Hi @chanetd as a new contributor, I'd like to recommend going forward that once we've begun a review discussion on changes in a PR, please refrain from squashing new commits into older commits.  This just makes the review process of newer changes easier since we can review the newer commits separate from what we have already reviewed in older commits.  If we want the commits to be squashed later we can request that after-the-fact or do it ourselves.  It's not a big deal this time, but it just makes the review/merge process easier for us going forward.

chanetd commented:
> Hi @chanetd as a new contributor, I'd like to recommend going forward that once we've begun a review discussion on changes in a PR, please refrain from squashing new commits into older commits. This just makes the review process of newer changes easier since we can review the newer commits separate from what we have already reviewed in older commits. If we want the commits to be squashed later we can request that after-the-fact or do it ourselves. It's not a big deal this time, but it just makes the review/merge process easier for us going forward.

Got it. What about rebasing against the tip of the `main` branch? Should I still do that before every push, or leave that off until the review has settled?

Naros commented:
If the PR is touching quite a lot of things, it's probably best to delay the rebase less we explicitly ask you to rebase in advance.  If we haven't yet reviewed any of the work, feel free to rebase and force push as needed.  If the PR changes are very tiny, I don't personally think it matters when you rebase.

Naros commented:
Merged to main with some changes (=1.9), will backport to 1.8 at 1.8.1.Final release time.
